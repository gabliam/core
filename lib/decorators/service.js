"use strict";
const constants_1 = require("../constants");
const container_1 = require("../container");
const registry_1 = require("../registry");
function Service(name) {
    return function (target) {
        let id = name ? name : target;
        container_1.fluentProvider(id)
            .inSingletonScope()
            .done()(target);
        registry_1.registry.add(constants_1.TYPE.Service, id);
    };
}
exports.Service = Service;
