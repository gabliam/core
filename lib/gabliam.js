"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
const express = require("express");
const constants_1 = require("./constants");
const loader_1 = require("./loader");
const container_1 = require("./container");
const registry_1 = require("./registry");
/**
 * Wrapper for the express server.
 */
class Gabliam {
    /**
     * Wrapper for the express server.
     *
     * @param container Container loaded with all controllers and their dependencies.
     */
    constructor(discoverPath, customRouter, routingConfig) {
        this.discoverPath = discoverPath;
        this._plugins = [];
        this.container = container_1.container;
        this._app = express();
        this._router = customRouter || express.Router();
        this._routingConfig = routingConfig || {
            rootPath: constants_1.DEFAULT_ROUTING_ROOT_PATH
        };
    }
    /**
     * Sets the configuration function to be applied to the application.
     * Note that the config function is not actually executed until a call to InversifyExpresServer.build().
     *
     * This method is chainable.
     *
     * @param fn Function in which app-level middleware can be registered.
     */
    setConfig(fn) {
        this._configFn = fn;
        return this;
    }
    /**
     * Sets the error handler configuration function to be applied to the application.
     * Note that the error config function is not actually executed until a call to InversifyExpresServer.build().
     *
     * This method is chainable.
     *
     * @param fn Function in which app-level error handlers can be registered.
     */
    setErrorConfig(fn) {
        this._errorConfigFn = fn;
        return this;
    }
    addPlugin(fn) {
        this._plugins.push(fn);
        return this;
    }
    /**
     * Applies all routes and configuration to the server, returning the express application.
     */
    build() {
        return __awaiter(this, void 0, void 0, function* () {
            loader_1.loader(this.discoverPath);
            this._loadConfig();
            yield Promise.all(this._plugins.map(plugin => plugin(this)));
            // register server-level middleware before anything else
            if (this._configFn) {
                this._configFn.apply(undefined, [this._app]);
            }
            this.registerControllers();
            // register error handlers after controllers
            if (this._errorConfigFn) {
                this._errorConfigFn.apply(undefined, [this._app]);
            }
            return this._app;
        });
    }
    _loadConfig() {
        let configIds = registry_1.registry.get(constants_1.TYPE.Config);
        configIds.forEach(configId => {
            let confInstance = this.container.get(configId);
            let methodMetadata = Reflect.getOwnMetadata(constants_1.METADATA_KEY.Bean, confInstance.constructor);
            if (methodMetadata) {
                methodMetadata.forEach(metadata => {
                    this.container.bind(metadata.id).toConstantValue(confInstance[metadata.key]());
                });
            }
        });
    }
    registerControllers() {
        let controllerIds = registry_1.registry.get(constants_1.TYPE.Controller);
        controllerIds.forEach((controllerId) => {
            let controller = this.container.get(controllerId);
            let controllerMetadata = Reflect.getOwnMetadata(constants_1.METADATA_KEY.controller, controller.constructor);
            let methodMetadata = Reflect.getOwnMetadata(constants_1.METADATA_KEY.controllerMethod, controller.constructor);
            if (controllerMetadata && methodMetadata) {
                methodMetadata.forEach((metadata) => {
                    console.log(`${controllerMetadata.path}${metadata.path}`);
                    let handler = this.handlerFactory(controllerId, metadata.key, controllerMetadata.json);
                    this._router[metadata.method](`${controllerMetadata.path}${metadata.path}`, ...controllerMetadata.middlewares, ...metadata.middlewares, handler);
                });
            }
        });
        this._app.use(this._routingConfig.rootPath, this._router);
    }
    handlerFactory(controllerId, key, json) {
        return (req, res, next) => {
            let result = this.container.get(controllerId)[key](req, res, next);
            // try to resolve promise
            if (result && result instanceof Promise) {
                result.then((value) => {
                    if (value && !res.headersSent) {
                        if (json) {
                            res.json(value);
                        }
                        else {
                            res.send(value);
                        }
                    }
                })
                    .catch((error) => {
                    next(error);
                });
            }
            else if (result && !res.headersSent) {
                if (json) {
                    res.json(result);
                }
                else {
                    res.send(result);
                }
            }
        };
    }
}
exports.Gabliam = Gabliam;
