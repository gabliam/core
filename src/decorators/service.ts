import { TYPE } from '../constants';
import { interfaces, injectable } from 'inversify';
import { register } from './register';

export function Service(name?: string) {
    return function (target: any) {
        let id: interfaces.ServiceIdentifier<any> = name ? name : target;
        injectable()(target);
        register(TYPE.Service, { id, target })(target);
    };
}
