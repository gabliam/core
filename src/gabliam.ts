import * as inversify from 'inversify';
import * as interfaces from './interfaces';
import * as _ from 'lodash';
import {
  TYPE,
  METADATA_KEY,
  APP_CONFIG,
  CORE_CONFIG,
  VALUE_EXTRACTOR
} from './constants';
import { Loader } from './loader';
import { createContainer } from './container';
import { Registry } from './registry';
import * as d from 'debug';
import { configureValueExtractor } from './utils';

const debug = d('Gabliam:core');

const DEFAULT_CONFIG: interfaces.GabliamConfig = {
  scanPath: process.env.PWD!,
  configPath: process.env.GABLIAM_CONFIG_PATH || process.env.PWD!
};

/**
 * Gabliam
 */
export class Gabliam {
  public container: inversify.interfaces.Container = createContainer();

  public config: any;

  /**
   * Registry
   */
  public registry: Registry;

  private _loader: Loader = new Loader();

  private _plugins: interfaces.GabliamPlugin[] = [];
  private _options: interfaces.GabliamConfig;

  /**
   * Constructor
   * @param  {interfaces.GabliamConfig|string} options?
   */
  constructor(options?: Partial<interfaces.GabliamConfig> | string) {
    if (options === undefined) {
      this._options = DEFAULT_CONFIG;
    } else {
      if (_.isString(options)) {
        this._options = {
          scanPath: options,
          configPath: options
        };
      } else {
        this._options = {
          ...DEFAULT_CONFIG,
          ...options
        };
      }
    }
    /**
     * @TODO move in building phase
     */
    this.container.bind(CORE_CONFIG).toConstantValue(this._options);
  }
  /**
   * Add a plugin order
   * @param  {interfaces.GabliamPluginConstructor} ctor
   * @returns Gabliam
   */
  public addPlugin(ctor: interfaces.GabliamPluginConstructor): Gabliam {
    this._plugins.push(new ctor());
    return this;
  }

  /**
   * Build gabliam
   * @returns Promise
   */
  public async build(): Promise<Gabliam> {
    /**
     * Load config file
     */
    this._initializeConfig();

    /**
     * Loading phase
     */
    this.registry = this._loader.loadModules(
      this._options.scanPath,
      this._plugins
    );

    /**
     * Binding phase
     */
    this._bind();

    /**
     * Config phase
     */
    await this._loadConfig();

    /**
     * building phase
     */
    this._plugins
      .filter(plugin => _.isFunction(plugin.build))
      .forEach(plugin => plugin.build!(this.container, this.registry));

    return this;
  }

  /**
   * Build and start gabliam application
   * @returns Promise
   */
  async buildAndStart(): Promise<Gabliam> {
    await this.build();
    await this.start();
    return this;
  }

  /**
   * Start gabliam application
   *
   * call all plugin.start
   * @returns Promise
   */
  async start(): Promise<Gabliam> {
    const pluginsStart = this._plugins
      .filter(plugin => _.isFunction(plugin.start))
      .map(plugin => plugin.start!(this.container, this.registry));

    await Promise.all(pluginsStart);
    return this;
  }

  /**
   * Stop gabliam application
   * call all plugin.stop
   *
   * @returns Promise
   */
  async stop(): Promise<Gabliam> {
    const pluginsStop = this._plugins
      .filter(plugin => _.isFunction(plugin.stop))
      .map(plugin => plugin.stop!(this.container, this.registry));

    await Promise.all(pluginsStop);
    return this;
  }

  /**
   * Destroy gabliam application
   * call all plugin.destroy
   *
   * @returns Promise
   */
  async destroy(): Promise<Gabliam> {
    const pluginsDestroy = this._plugins
      .filter(plugin => _.isFunction(plugin.destroy))
      .map(plugin => plugin.destroy!(this.container, this.registry));

    await Promise.all(pluginsDestroy);
    return this;
  }

  /**
   * Load config file and bind result in APP_CONFIG
   */
  private _initializeConfig() {
    const config = (this.config = this._loader.loadConfig(
      this._options.configPath
    ));
    this.container.bind(APP_CONFIG).toConstantValue(config);
    this.container
      .bind(VALUE_EXTRACTOR)
      .toConstantValue(configureValueExtractor(this.container));
  }

  /**
   * Binding phase
   * Binding of config and service classes.
   * call all plugin.bind
   */
  private _bind() {
    debug('_bind');
    this.registry
      .get(TYPE.Config)
      .forEach(({ id, target }) =>
        this.container.bind(id).to(target).inSingletonScope()
      );

    this.registry
      .get(TYPE.Service)
      .forEach(({ id, target }) =>
        this.container.bind(id).to(target).inSingletonScope()
      );

    this._plugins
      .filter(plugin => _.isFunction(plugin.bind))
      .forEach(plugin => plugin.bind!(this.container, this.registry));
  }

  /**
   * Config phase
   */
  private async _loadConfig() {
    debug('_loadConfig');
    async function callInstance(instance: any, key: string) {
      return Promise.resolve(instance[key]());
    }

    // list of plugins with config phase
    const pluginConfig = this._plugins.filter(plugin =>
      _.isFunction(plugin.config)
    );

    let configsRegistry = this.registry.get<interfaces.ConfigRegistry>(
      TYPE.Config
    );

    debug('configsRegistry', configsRegistry);
    if (configsRegistry) {
      configsRegistry = _.sortBy(configsRegistry, 'order');

      for (const { id: configId } of configsRegistry) {
        const confInstance = this.container.get<object>(configId);

        const beanMetadata: interfaces.BeanMetadata[] = Reflect.getOwnMetadata(
          METADATA_KEY.bean,
          confInstance.constructor
        );

        if (beanMetadata) {
          // No promise.all and await because order of beans are important
          for (const metadata of beanMetadata) {
            this.container
              .bind(metadata.id)
              .toConstantValue(await callInstance(confInstance, metadata.key));
          }
        }

        pluginConfig.forEach(plugin =>
          plugin.config!(this.container, this.registry, confInstance)
        );
      }
    }
    debug('_loadConfig end');
  }
}
