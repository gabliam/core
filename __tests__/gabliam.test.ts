// tslint:disable:one-line
// tslint:disable:no-unused-expression

import {
  Gabliam,
  APP_CONFIG,
  CORE_CONFIG,
  interfaces,
  inversifyInterfaces,
  Registry,
  Config,
  Plugin
} from '../src';
import * as path from 'path';
import { TestService } from './fixtures/gabliam/service';
import { DbConfig } from './fixtures/gabliam/db-config';
import * as sinon from 'sinon';
import { GabliamTest } from '../src/testing/gabliam';

test('gabliam instance', async () => {
  const gab = new Gabliam({
    scanPath: path.resolve(__dirname, './fixtures/gabliam'),
    configPath: path.resolve(__dirname, './fixtures/gabliam/config')
  });
  await gab.buildAndStart();
  // @todo write a guid serializer
  // expect(gab).toMatchSnapshot();
  expect(gab.container.get(APP_CONFIG)).toMatchSnapshot();
  expect(gab.container.get(CORE_CONFIG)).toMatchSnapshot();
  expect(gab.container.get(TestService)).toMatchSnapshot();
  expect(gab.container.get(DbConfig)).toMatchSnapshot();
  await gab.destroy();
});

test('gabliam instance with default config', async () => {
  const gab = new Gabliam();
  expect(gab.options).toEqual({
    scanPath: process.env.PWD,
    configPath: process.env.PWD
  });
  await gab.destroy();
});

test('gabliam instance with path', async () => {
  const gab = new Gabliam(path.resolve(__dirname, './fixtures/gabliam'));
  await gab.buildAndStart();
  // @todo write a guid serializer
  // expect(gab).toMatchSnapshot();
  expect(gab.container.get(APP_CONFIG)).toMatchSnapshot();
  expect(gab.container.get(CORE_CONFIG)).toMatchSnapshot();
  expect(gab.container.get(TestService)).toMatchSnapshot();
  expect(gab.container.get(DbConfig)).toMatchSnapshot();
  await gab.destroy();
});

describe('test plugin', async () => {
  @Plugin()
  class PluginTest implements interfaces.GabliamPlugin {
    build(container: inversifyInterfaces.Container, registry: Registry): void {}

    bind(container: inversifyInterfaces.Container, registry: Registry): void {}

    config(
      container: inversifyInterfaces.Container,
      registry: Registry,
      confInstance: any
    ): void {}

    start(
      container: inversifyInterfaces.Container,
      registry: Registry
    ): Promise<void> {
      return Promise.resolve();
    }

    stop(
      container: inversifyInterfaces.Container,
      registry: Registry
    ): Promise<void> {
      return Promise.resolve();
    }

    destroy(
      container: inversifyInterfaces.Container,
      registry: Registry
    ): Promise<void> {
      return Promise.resolve();
    }
  }

  const build = sinon.spy(PluginTest.prototype, 'build');
  const bind = sinon.spy(PluginTest.prototype, 'bind');
  const config = sinon.spy(PluginTest.prototype, 'config');
  const start = sinon.spy(PluginTest.prototype, 'start');
  const stop = sinon.spy(PluginTest.prototype, 'stop');
  const destroy = sinon.spy(PluginTest.prototype, 'destroy');
  let gab: Gabliam;

  describe('with config', () => {
    let res = '';
    beforeAll(() => {
      const g = new GabliamTest(
        new Gabliam({
          scanPath: path.resolve(__dirname, 'gabliam'),
          configPath: path.resolve(__dirname, 'gabliam')
        }).addPlugin(PluginTest)
      );
      gab = g.gab;

      @Config(300)
      class Conf {
        constructor() {
          res += 'Conf';
        }
      }
      @Config(200)
      class Conf2 {
        constructor() {
          res += 'Conf2';
        }
      }

      g.addClass(Conf);
      g.addClass(Conf2);
    });

    afterAll(() => {
      build.reset();
      bind.reset();
      config.reset();
      start.reset();
      stop.reset();
      destroy.reset();
    });

    test('gabliam build', async () => {
      await gab.build();
      expect(build.calledOnce).toBe(true);
      expect(bind.calledOnce).toBe(true);
      expect(config.callCount).toMatchSnapshot();
      expect(res).toMatchSnapshot();
    });

    test('gabliam start', async () => {
      await gab.start();
      expect(start.calledOnce).toBe(true);
    });

    test('gabliam stop', async () => {
      await gab.stop();
      expect(stop.calledOnce).toBe(true);
    });

    test('gabliam destroy', async () => {
      await gab.destroy();
      expect(destroy.calledOnce).toBe(true);
    });
  });

  describe('without config', () => {
    beforeAll(() => {
      gab = new Gabliam({
        scanPath: path.resolve(__dirname, './fixtures/gabliam'),
        configPath: path.resolve(__dirname, './fixtures/gabliam/config')
      });
      gab.addPlugin(PluginTest);
    });

    afterAll(() => {
      build.reset();
      bind.reset();
      config.reset();
      start.reset();
      stop.reset();
      destroy.reset();
    });

    test('gabliam build', async () => {
      await gab.build();
      expect(build.calledOnce).toBe(true);
      expect(bind.calledOnce).toBe(true);
      expect(config.calledOnce).toBe(true);
    });

    test('gabliam start', async () => {
      await gab.start();
      expect(start.calledOnce).toBe(true);
    });

    test('gabliam stop', async () => {
      await gab.stop();
      expect(stop.calledOnce).toBe(true);
    });

    test('gabliam destroy', async () => {
      await gab.destroy();
      expect(destroy.calledOnce).toBe(true);
    });
  });
});

test('should fail with bad Plugin', () => {
  class BadPlugin {}
  const g = new GabliamTest();
  expect(() => {
    g.gab.addPlugin(BadPlugin);
  }).toThrowError();
});

test('should fail with Plugin with bad deps', () => {
  @Plugin({ dependencies: ['BadPlugin'] })
  class BadPlugin {}
  const g = new GabliamTest();
  expect(() => {
    g.gab.addPlugin(BadPlugin);
  }).toThrowError();
});

test('plugin findByName', () => {
  @Plugin()
  class BadPlugin {}
  const g = new GabliamTest();
  g.gab.addPlugin(BadPlugin);
  expect(g.gab.pluginList.findByName('BadPlugin')).toMatchSnapshot();
});
