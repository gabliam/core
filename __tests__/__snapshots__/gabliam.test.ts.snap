// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gabliam instance 1`] = `
Object {
  "application": Object {
    "db": "test",
    "host": "127.0.0.1",
  },
}
`;

exports[`gabliam instance 2`] = `
Object {
  "configPath": "<PROJECT_ROOT>/__tests__/fixtures/gabliam/config",
  "scanPath": "<PROJECT_ROOT>/__tests__/fixtures/gabliam",
}
`;

exports[`gabliam instance 3`] = `
TestService {
  "dbConfig": DbConfig {
    "db": "test",
    "host": "127.0.0.1",
  },
}
`;

exports[`gabliam instance 4`] = `
DbConfig {
  "db": "test",
  "host": "127.0.0.1",
}
`;

exports[`gabliam instance with path 1`] = `Object {}`;

exports[`gabliam instance with path 2`] = `
Object {
  "configPath": "<PROJECT_ROOT>/__tests__/fixtures/gabliam",
  "scanPath": "<PROJECT_ROOT>/__tests__/fixtures/gabliam",
}
`;

exports[`gabliam instance with path 3`] = `
TestService {
  "dbConfig": DbConfig {
    "db": undefined,
    "host": undefined,
  },
}
`;

exports[`gabliam instance with path 4`] = `
DbConfig {
  "db": undefined,
  "host": undefined,
}
`;

exports[`test plugin with config gabliam build 1`] = `2`;

exports[`test plugin with config gabliam build 2`] = `"Conf2Conf"`;
